File: drive_sphero.go
// Call on this program [must be in same directory] through: 
// $ go build drive_sphero.go // builds the program based on latest saved version
// $ sudo ./drive_sphero SK_2AE2 // runs program and passes the correct SPRK+ address

package main

import (
	"fmt"
	"os"
        "time"

	"gobot.io/x/gobot" // gobot framework
	
	"github.com/eiannone/keyboard" // for keyboard commands
	
        "gobot.io/x/gobot/platforms/ble" // connecting through BLE
        "gobot.io/x/gobot/platforms/sphero/sprkplus" // SPRK specific commands
)

/*
https://github.com/hybridgroup/gobot/blob/master/platforms/sphero/sphero_driver.go
^^ Website for the sphero driver various function calls
*/

func main() {
	bleAdaptor := ble.NewClientAdaptor(os.Args[1]) // gets passed when program is called
        sprk := sprkplus.NewDriver(bleAdaptor)
	
	// setting color variables as max on each
	r := uint8(255)
      	g := uint8(255)
     	b := uint8(255)
      	
	work := func() {
		fmt.Printf("Opening Keyboard\n")
		if err := keyboard.Open(); err != nil { // open the keyboard
			panic(err)
		}
		fmt.Printf("Keyboard is open\n")			
		var exit bool	
		fmt.Printf("Press Ctrl+C to exit at any time\n")
		fmt.Printf("Use the Arrows to Control Sphero\n")
			
		for !exit { // will run infinitely until CTRL+C is pressed
			char, key, err := keyboard.GetKey() // always searching for keys
			//keysEvent, err := keyboard.GetKeys(1)
			if err != nil {
				panic(err)
			}
			//key := <-keysEvent
			switch {
			case key == keyboard.KeyArrowUp:
				fmt.Printf("You pressed Up\t\t\tDriving Forward\n")
				sprk.SetRGB(r, 0, 0) // RED
				sprk.Roll(50,0) // 100 speed, heading forward
			case key == keyboard.KeyArrowRight:
				fmt.Printf("You pressed Right\t\tDriving Right\n")
				sprk.SetRGB(0, 0, b) // BLUE 
				sprk.Roll(50,90) // 100 speed, heading right
			case key == keyboard.KeyArrowLeft:
				fmt.Printf("You pressed Left\t\tDriving Left\n")
				sprk.SetRGB(0, g, 0) // GREEN
				sprk.Roll(50,270) // 100 speed, heading left
			case key == keyboard.KeyArrowDown:
				fmt.Printf("You pressed Down\t\tDriving Backwards\n")
				sprk.SetRGB(r, g, b) // WHITE
				sprk.Roll(50,180) // 100 speed, heading backwards
			case key == keyboard.KeyCtrlC:
				fmt.Printf("You want to Exit\n")
				exit = true // loop will no longer continue, will exit
			default:
				fmt.Printf("You pressed ", char)
			}
			time.Sleep((1)*time.Second) // wait the time before next loop
			sprk.Stop() // stop rolling
		}
		fmt.Printf("Proceeding to Exit...\n") // exits when CTRL+C is pressed
		_ = keyboard.Close() // close keyboard
		fmt.Printf("Keyboard is Closed\nPress Ctrl+C again to Exit Program\n")
	}
	
	robot := gobot.NewRobot("sprkBot",
                []gobot.Connection{bleAdaptor},
                []gobot.Device{sprk},
                work,
        )

	robot.Start()
}

File: drive_sphero.js
// Functions of the sphero:
// https://github.com/sphero-inc/sphero.js/blob/e9010264a7f8fdca46fb304eb9550b96de036626/lib/devices/sphero.js#L433

var mac_id = "C5:9C:75:02:2A:E2"; // BLE ID for the specific Sphero SPRK+ (SK-2AE2)
var sphero = require( "sphero" ),
	sprk = sphero( mac_id ); 

console.log( 'Connecting with Sphero SPRK+ ...' );
sprk.connect(); // connect the robot (light should turn on)
console.log( mac_id, 'is successfully connected\n\nProceeding to start' );
arrowKey_control();
console.log( 'The Sphero is ready to drive:\n\t\tUse the arrow keys for manual control' );
console.log( '\t\tPress [b] to toggle the rear light for direction' );
console.log( '\t\tPress [Ctrl+c] at any time to quit the program\n' );

function arrowKey_control() {		
	var Keys = { left: '\u001b[D', up: '\u001b[A', right: '\u001b[C', down: '\u001b[B' }; 
	//Telnet keyboard equivelant for arrows
	var toggle = true; // bool value for toggling rear light
	var direction = 'start'; // values to keep track of latest direction
	var dir_history = []; // array to keep track of past directions and limits
	var dir_limit = 'none'; // used as a 4-way bool to determine a direction limit
	var limit_value = 4; // the value of how many times you can travel the same direction
	var time_reg = 400; // ms
	var time_180 = 700; // ms
	var sphero_speed = 80; // speed of the sphero, used in roll functions
	
	var stdin = process.stdin // look for keystrokes in terminal window
	stdin.setRawMode( true ); // registers continuous typing (no 'enter' needed)
	stdin.resume(); // process wil not exit
	stdin.setEncoding( 'ascii' ); // encoding for recieved data will be readable
	
	stdin.on( 'data', function( key ) {
		sprk.setRgbLed({ red: 0, green: 0, blue: 0 }); // turn off the main LED's

		if( key === '\u0003' ) { // key code for CTRL+C
			console.log( '\nExitting Process...\nEnsure the Sphero is off before reuse\n' );				
			process.exit() // end the program
		}
		else if( key === 'b' ) { 
			if( toggle ) {
				sprk.setBackLed( 255 ); // press 'b' to show back light
				toggle = false;
			}
			else {
				sprk.setBackLed( 0 ); // press again to turn off light
				toggle = true;
			}
		}
		else if( key === Keys.left && dir_limit != 'left' ) { // if left is clicked and hasn't been pressed 'limit_value' times before
			if( direction === 'right' ) // if opposite direction, needs more time to rotate and move
				sprk.setMotionTimeout( time_180 );
			else 
				sprk.setMotionTimeout( time_reg ); // standard time, only rolls for 400ms
				
			sprk.roll( sphero_speed, 270 ); // roll( speed, heading )
			console.log( '\tDriving Left' ); // print out the direction moved
			direction = 'left'; // logs the latest direction moved
			dir_limit = 'none' // reset the direction limit
		}
		else if( key === Keys.right && dir_limit != 'right' ) {
			if( direction === 'left' )
				sprk.setMotionTimeout( time_180 );
			else 
				sprk.setMotionTimeout( time_reg );
			
			sprk.roll( sphero_speed, 90 );	
			console.log( '\tDriving Right' );	
			direction = 'right';
			dir_limit = 'none'	
		}
		else if( key === Keys.up && dir_limit != 'up' ) {
			if( direction === 'down' ) 
				sprk.setMotionTimeout( time_180 )
			else 
				sprk.setMotionTimeout( time_reg );
			
			sprk.roll( sphero_speed, 0 );
			console.log( '\tDriving Forwards' );
			direction = 'up';
			dir_limit = 'none'	
		}
		else if( key === Keys.down && dir_limit != 'down' ) {
			if( direction === 'up' )
				sprk.setMotionTimeout( time_180 );
			else
				sprk.setMotionTimeout( time_reg );
				
			sprk.roll( sphero_speed, 180 );	
			console.log( '\tDriving Backwards' );
			direction = 'down';
			dir_limit = 'none'	
		}
		
		sprk.setStabilization( 1 );
		dir_history[ dir_history.length ] = direction; // update the array with the latest direction
		
		if( dir_history.length > limit_value-1 ) { // if there are more values in the array then the desired limit
			var last_dir = dir_history[ dir_history.length-1 ]; // variable for the last direction (could also use 'direction')
			var limit = true; // boolean to keep track of if the past 'limit_value' directions were the same
		
			for( i = dir_history.length-1; i > dir_history.length-limit_value-1 && limit; i-- ) {
				if( last_dir != dir_history[ i ] )
					limit = false; 
			// for loop will keep limit true if the past 'limit_value' directions were the same, else will be false
			}
			if( limit )
					dir_limit = last_dir; // limit is set to the direction that was repeated 'limit_value' times
		}
	});
}

